sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/variant/variantExpressions.scala

```diff
@@ -53,3 +66,311 @@ case class ParseJson(child: Expression)
+    for (path <- parsedPath) {
+      v = path match {
+        case scala.util.Left(key) if v.getType == Type.OBJECT => v.getFieldByKey(key)
+        case scala.util.Right(index) if v.getType == Type.ARRAY => v.getElementAtIndex(index)
+        case _ => null
+      }
+      if (v == null) return null
+    }
+    VariantGet.cast(v, dataType, failOnError, zoneId)
+  }
+
+  /**
+   * Cast a variant `v` into a target data type `dataType`. If the variant represents a variant
+   * null, the result is always a SQL NULL. The cast may fail due to an illegal type combination
+   * (e.g., cast a variant int to binary), or an invalid input valid (e.g, cast a variant string
+   * "hello" to int). If the cast fails, throw an exception when `failOnError` is true, or return a
+   * SQL NULL when it is false.
+   */
+  def cast(v: Variant, dataType: DataType, failOnError: Boolean, zoneId: Option[String]): Any = {
+    def invalidCast(): Any =
+      if (failOnError) throw QueryExecutionErrors.invalidVariantCast(v.toJson, dataType) else null
+
+    val variantType = v.getType
+    if (variantType == Type.NULL) return null
+    dataType match {
+      case VariantType => new VariantVal(v.getValue, v.getMetadata)
+      case _: AtomicType =>
+        variantType match {
+          case Type.OBJECT | Type.ARRAY =>
+            if (dataType == StringType) UTF8String.fromString(v.toJson) else invalidCast()
+          case _ =>
+            val input = variantType match {
+              case Type.BOOLEAN => v.getBoolean
+              case Type.LONG => v.getLong
+              case Type.STRING => UTF8String.fromString(v.getString)
+              case Type.DOUBLE => v.getDouble
+              case Type.DECIMAL => Decimal(v.getDecimal)
+              // We have handled other cases and should never reach here. This case is only intended
+              // to by pass the compiler exhaustiveness check.
+              case _ => throw QueryExecutionErrors.unreachableError()
+            }
+            // We mostly use the `Cast` expression to implement the cast. However, `Cast` silently
+            // ignores the overflow in the long/decimal -> timestamp cast, and we want to enforce
+            // strict overflow checks.
+            input match {
+              case l: Long if dataType == TimestampType =>
+                try Math.multiplyExact(l, MICROS_PER_SECOND)
+                catch {
+                  case _: ArithmeticException => invalidCast()
+                }
+              case d: Decimal if dataType == TimestampType =>
+                try {
+                  d.toJavaBigDecimal
+                    .multiply(new java.math.BigDecimal(MICROS_PER_SECOND))
+                    .toBigInteger
+                    .longValueExact()
+                } catch {
+                  case _: ArithmeticException => invalidCast()
+                }
+              case _ =>
+                val result = Cast(Literal(input), dataType, zoneId, EvalMode.TRY).eval()
```

Comment:
It's risky to evaluate Cast on the fly, as we do not apply any analysis checks. Can we define the allowed type mapping before creating `Cast`.

Reply:
I don't think so. The target type can be boolean and the variant value can be int type, which isn't a valid cast in the ANSI mode, but we won't see any failure if we evaluate the Cast expression directly.

I suggest we check `Cast.canAnsiCast` and fail earlier.